aws-api/src/main/java/com/amplifyframework/api/aws/auth/AuthRuleProcessor.java

```diff
@@ -0,0 +1,222 @@
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amplifyframework.api.aws.auth;
+
+import androidx.annotation.NonNull;
+
+import com.amplifyframework.AmplifyException;
+import com.amplifyframework.api.ApiException;
+import com.amplifyframework.api.aws.ApiAuthProviders;
+import com.amplifyframework.api.aws.AppSyncGraphQLRequest;
+import com.amplifyframework.api.aws.AuthorizationType;
+import com.amplifyframework.api.aws.sigv4.CognitoUserPoolsAuthProvider;
+import com.amplifyframework.api.aws.sigv4.DefaultCognitoUserPoolsAuthProvider;
+import com.amplifyframework.api.aws.sigv4.OidcAuthProvider;
+import com.amplifyframework.api.graphql.GraphQLRequest;
+import com.amplifyframework.core.model.AuthRule;
+import com.amplifyframework.core.model.AuthStrategy;
+import com.amplifyframework.core.model.ModelOperation;
+
+import com.amazonaws.mobileconnectors.cognitoidentityprovider.exceptions.CognitoParameterInvalidException;
+import com.amazonaws.mobileconnectors.cognitoidentityprovider.util.CognitoJWTParser;
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Objects;
+
+/**
+ * Processor that can decorate an AppSync-compliant GraphQL request with additional variables
+ * that are required for owner-based or group-based authorization.
+ */
+public final class AuthRuleProcessor {
+    private final ApiAuthProviders authProvider;
+
+    /**
+     * Constructs a new instance of GraphQL request's auth rule processor.
+     * @param authProvider the auth providers to authorize requests
+     */
+    public AuthRuleProcessor(@NonNull ApiAuthProviders authProvider) {
+        this.authProvider = Objects.requireNonNull(authProvider);
+    }
+
+    /**
+     * Decorate given GraphQL request instance with additional variables for owner-based or
+     * group-based authorization.
+     *
+     * This will only work if the request is compliant with the AppSync specifications.
+     * @param request an instance of {@link GraphQLRequest}.
+     * @param authType the mode of authorization being used to authorize the request
+     * @param <R> The type of data contained in the GraphQLResponse expected from this request.
+     * @return the input request with additional variables that specify model's owner and/or
+     *          groups
+     * @throws ApiException If an error is encountered while processing the auth rules associated
+     *          with the request or if the authorization fails
+     */
+    public <R> GraphQLRequest<R> process(
```

Comment:
maybe 
```suggestion
    public <R> GraphQLRequest<R> addAuthorizationFields(
```
or: 
```suggestion
    public <R> GraphQLRequest<R> decorate(
```
?

Should this be a `AuthRuleRequestDecorator`?